test/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -40,82 +40,126 @@ class SingleSignOnProfileVerifier(val response: Node) {
+                throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLProfiles_4_1_4_2_c,
+                        "Format",
+                        issuerFormat,
+                        "urn:oasis:names:tc:SAML:2.0:nameid-format:entity")
         }
     }
 
     /**
      * Verify Assertion Elements against the profile document
      * 4.1.4.2 <Response> Usage
      */
-    fun verifySsoAssertions() {
+    private fun verifySsoAssertions() {
         val assertions = response.children("Assertion")
-        if (assertions.isEmpty()) throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_d, message = "No Assertions found.")
+        if (assertions.isEmpty()) {
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_d, message = "No Assertions found.")
+        }
 
         assertions.forEach {
             verifyIssuer()
-
-            if (it.children("Subject").size != 1) throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_f, message = "${it.children("Subject").size} Subject elements were found.")
-
-            val bearerSubjectConfirmations = mutableListOf<Node>()
-            it.children("Subject")[0].children("SubjectConfirmation")
-                    .filter { it.attributes.getNamedItem("Method").textContent == "urn:oasis:names:tc:SAML:2.0:cm:bearer" }
-                    .toCollection(bearerSubjectConfirmations)
-
-            //todo - We can't throw an exception here because the spec says there could be assertions without bearer SubjectConfirmation
-            if (bearerSubjectConfirmations.isEmpty())
-                throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_g, message = "No bearer SubjectConfirmation elements were found.")
-
-            // Check if there is one SubjectConfirmationData with a Recipient, InResponseTo and NotOnOrAfter
-            if (bearerSubjectConfirmations
-                            .filter { it.children("SubjectConfirmationData").isNotEmpty() }
-                            .flatMap { it.children("SubjectConfirmationData") }
-                            .none {
-                                it.attributes.getNamedItem("Recipient").textContent == ACS_URL &&
-                                        it.attributes.getNamedItem("InResponseTo").textContent == ID &&
-                                        it.attributes.getNamedItem("NotOnOrAfter") != null &&
-                                        it.attributes.getNamedItem("NotBefore") == null
-                            })
-                throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_h, message = "There were no bearer SubjectConfirmation elements that matched the criteria below.")
+            verifyHasSubject(it)
+            verifyBearerSubjectConfirmations(it)
 
             //todo - We can't throw an exception here either
```

Comment:
✏️ Could we put `because the spec says there could be assertions without bearer` instead of the word `either` now that the context of the previous comments is in a different place?

Reply:
yes